webapp/components/status_dropdown.jsx

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import React from 'react';
+import {Dropdown} from 'react-bootstrap';
+import StatusIcon from './status_icon.jsx';
+import * as ChannelActions from 'actions/channel_actions.jsx';
+import ChannelStore from 'stores/channel_store.jsx';
+import TeamStore from 'stores/team_store.jsx';
+import {FormattedMessage} from 'react-intl';
+import PreferenceStore from 'stores/preference_store.jsx';
+
+const STATUSES = ['online', 'away', 'offline'];
+
+export default class StatusDropdown extends React.Component {
+
+    static propTypes = {
+        status: React.PropTypes.string
+    }
+
+    constructor(props) {
+        super(props);
+
+        [
+            'toggleDropdown',
+            'setStatus',
+            'renderStatusAction',
+            'onMouseEnter',
+            'onMouseLeave'
+        ].forEach((fnName) => {
+            this[fnName] = this[fnName].bind(this);
+        });
+
+        this.state = {
+            showDropdown: false,
+            mouseOver: false
+        };
+    }
+
+    onMouseEnter() {
+        this.setState({mouseOver: true});
+    }
+
+    onMouseLeave() {
+        this.setState({mouseOver: false});
+    }
+
+    renderStatusIcon() {
+        const team = TeamStore.getCurrent();
+        const theme = PreferenceStore.getTheme(team.id);
+
+        if (this.state.mouseOver) {
+            return (
+                <span
+                    className={'status'}
+                    style={{color: theme.sidebarHeaderTextColor}}
+                >
+                    <i
+                        className={'fa fa-edit'}
+                        style={{padding: '1px 2px'}}
+                    />
+                </span>
+            );
+        }
+        return (
+            <StatusIcon
+                status={this.props.status}
+            />
+        );
+    }
+
+    toggleDropdown() {
+        const newState = {showDropdown: !this.state.showDropdown};
+        this.setState(newState);
+    }
+
+    setStatus = (status) => (event) => {
```

Comment:
Instead of having `setStatus` return a function, can you just make explicit `setStatusOnline`, `setStatusAway`, and `setStatusOnline` functions that just call some shared `setStatus(status)` to do the actual action? The way it is right now, we're creating new functions each time the component is rendered which is a bit wasteful. I think there's an ESLint warning when that happens currently